cassandra/protocol.py

```diff
@@ -122,32 +127,40 @@ def to_exception(self):
         return self
 
     class Codec(object):
-        opcode = 0x00
         error_decoders = None
 
         def __init__(self, error_decoders):
             self.error_decoders = {opcode: decoder for opcode, decoder in error_decoders}
 
-        def decode(self, f, protocol_version, *args, **kwargs):
+        def decode(self, f, protocol_version, *args):
             code = read_int(f)
             msg = read_string(f)
             error_decoder = self.error_decoders.get(code)
             if error_decoder:
-                return error_decoder(f, protocol_version, msg, **kwargs)
+                return error_decoder.decode(f, protocol_version, msg)
             return ErrorMessage(code=code, message=msg, info=None)
 
 
+class ErrorMessageSubTypeWithCodecMeta(type):
+    def __init__(cls, name, bases, clsdict):
+        super(ErrorMessageSubTypeWithCodecMeta, cls).__init__(name, bases, clsdict)
+        cls.Codec = type("Codec", (cls.Codec,), {})
```

Comment:
@mambocab  This is the part I had to add to the metaclass... otherwise all subclassing was overriding the error_class of ErrorMessageSub.Codec (those that simply reuse the default one). I don't especially like that... do you see a workaround?

Reply:
Ohh interesting, whoops. Would this be better?

```python
# Make unique subclass of the superclass's `.Codec` if necessary
# e.g. if you're defining `class FooMessage(ErrorMessageSub)` without its own
# `.Codec`, this will subclass `ErrorMessageSub.Codec` so that
# `FooMessage.Codec` and `ErrorMessageSub.Codec` are different objects.
if 'Codec' not in clsdict:
    class Codec(cls.Codec):
        pass
    clsdict['Codec'] = Codec

super(ErrorMessageSubTypeWithCodecMeta, cls).__init__(name, bases, clsdict)
# If we hadn't created a Codec subclass above, we could be altering a `Codec` that
# `cls` inherited
cls.Codec.error_class = cls
```

It's not quite as succinct, but I think it's clearer. The `if 'Codec' not in clsdict` also prevents us from creating unnecessary new subclasses. Another option:

```python
super(ErrorMessageSubTypeWithCodecMeta, cls).__init__(name, bases, clsdict)

if 'Codec' not in clsdict:
    # Make unique subclass of the superclass's `.Codec` if necessary
    # e.g. if you're defining `class FooMessage(ErrorMessageSub)` without its own
    # `.Codec`, this will subclass `ErrorMessageSub.Codec` so that
    # `FooMessage.Codec` and `ErrorMessageSub.Codec` are different objects.
    class Codec(cls.Codec):
        error_class = cls
    cls.Codec = Codec
else:
    cls.Codec.error_class = cls
```

I don't know off the top of my head if these have different behavior or produce observably different classes, but I think I slightly prefer the first option -- all `Codec` classes produced this way will be assigned to `cls.Codec` in `super(...).__init__`, and then have `error_class` assigned after creation.